### PR TITLE
add upload asset command

### PR DESCRIPTION
**Pull request by alenajk**
## Status
- [x] **Ready**
- [ ] **In progress**
- [ ] **Hold**

## Description
A few sentences describing the overall goals of the pull request's commits, as well as impacted areas or components of the library (e.g., Python API or CLI).

## Related PRs
- [x] This PR is independent

## TODOs
Add any todos here as needed for work in progress code. This section can be deleted if not relevant. Examples:

- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: onecodex/cli.py [hunk 1]
```diff
@@ -492,6 +492,40 @@ def upload(
...
+    "file",
+    nargs=1,
+    required=True,
+    type=click.Path(exists=True),
+    shell_complete=partial(click_path_autocomplete_helper, directory=False),
+)
+@click.pass_context
+@pretty_errors
+@telemetry
+@login_required(experimental=True)
```
Comment (petraszd) [inline onecodex/cli.py:512]: Nitpick: It is not clear what does `experimental` stand for. It seems that it tell to use an experimental login service or something.

Maybe  `@login_required(experimental_api=True)` or even more generic
`@login_required(api_kwargs={"experimental": True})`.

File: onecodex/lib/upload.py [hunk 2]
```diff
@@ -366,10 +364,56 @@ def _upload_document_fileobj(file_obj, file_name, documents_resource):
...
+        file_name,
+        fields,
+        assets_resource._client.session,
+        assets_resource._client._root_url + fields["callback_url"],  # full callback url
+    )
+
+    msg = f"{file_name}: finished"
+    document_id = s3_upload.get("document_id")
+    if document_id is not None:
+        msg += f" as document {document_id}"
```
Comment (pkonieczny) [inline onecodex/lib/upload.py:416]: Shouldn't it be `asset_id`?

File: onecodex/cli.py [hunk 3]
```diff
@@ -169,7 +169,7 @@ def _size_formatter(size):
 @click.pass_context
 @pretty_errors
 @telemetry
-@login_required
+@login_required()
```
Comment (pkonieczny) [inline onecodex/cli.py:172]: This looks a bit awkward. Maybe the new parametrised wrapper could be called `login_required_args`
and then have `login_required = partial(login_required_args, experimental=False)` and `login_required_experimental_api = partial(login_required_args, experimental=True)`?

**Review by pkonieczny — APPROVED** [review]
Some nits, but looks good overall!

File: onecodex/lib/upload.py (hunk 2 above)
Comment (alenajk) [reply onecodex/lib/upload.py:416]: Ah, actually `asset_id` isn't in the response.

File: onecodex/auth.py [hunk 4]
```diff
@@ -176,47 +176,58 @@ def _logout(creds_file=None):
...
-        elif api_key_env is not None:
-            api_kwargs["api_key"] = api_key_env
-        elif api_key_creds_file is not None:
-            api_kwargs["api_key"] = api_key_creds_file
-        else:
-            click.echo(
-                "The command you specified requires authentication. Please login first.\n", err=True
-            )
-            ctx.exit(1)
+def login_required_with_args(experimental=False):
```
Comment (petraszd) [inline onecodex/auth.py:179]: Nitpick: I would rename `experimental` to `experimental_api`. Now it seems that it is going to turn on "Experimental Login" functionality.

**Review by petraszd — APPROVED** [review]

File: onecodex/cli.py [hunk 5]
```diff
@@ -492,6 +498,40 @@ def upload(
     )
 
 
+@onecodex.command("upload_asset")
```
Comment (pkonieczny) [inline onecodex/cli.py:501]: For consistency, there should probably be an `asset` command group with `upload` command put inside, e.g.
`onecodex assets upload` vs `onecodex upload_asset`

Comment (alenajk) [reply onecodex/cli.py:501]: Good point - it's already a little inconsistent with `upload` (for FASTA/FASTQ) and then `documents upload`, but I agree having `assets upload` would be better. Will update.

**Review by pkonieczny — APPROVED** [review]

**Pull request merged**